packages/asana/src/Adaptor.js

```diff
@@ -233,39 +168,122 @@ export function createTask(params, callback) {
  */
 export function upsertTask(projectGid, params, callback) {
   return state => {
-    const resolvedProjectGid = expandReferences(projectGid)(state);
-    const { externalId, data } = expandReferences(params)(state);
-
-    const { apiVersion, token } = state.configuration;
-
-    const url = `https://app.asana.com/api/${apiVersion}/projects/${resolvedProjectGid}/tasks`;
-
-    const config = {
-      url,
-      headers: { Authorization: `Bearer ${token}` },
-      params: {
-        opt_fields: `${externalId}`,
-      },
-    };
-
-    return http
-      .get(config)(state)
-      .then(response => {
-        const matchingTask = response.data.data.find(
+    const [resolvedProjectGid, resolvedParams] = expandReferences(
+      state,
+      projectGid,
+      params
+    );
+    const { externalId, data } = resolvedParams;
+    return commonRequest(
+      state,
+      `projects/${resolvedProjectGid}/tasks`,
+      { query: { opt_fields: `${externalId}` } },
+      next => {
+        const matchingTask = next.data.find(
           task => task[externalId] === data[externalId]
         );
         if (matchingTask) {
           console.log('Matching task found. Performing update.');
           console.log('Data to update', data);
-          // projects and workspace ids should ne be included to update
-          delete data.projects;
-          delete data.workspace;
-          return updateTask(matchingTask.gid, data, callback)(state);
+          // projects and workspace ids should not be included to update
+          const { projects, workspace, ...remainingData } = data;
+          return updateTask(matchingTask.gid, remainingData, callback)(state);
         } else {
           console.log('No matching task found. Performing create.');
           return createTask(data, callback)(state);
         }
-      });
+      }
+    );
+  };
+}
+
+/**
+ * Options provided to the createTaskStory request
+ * @typedef {Object} StoryOptions
+ * @property {object} story - The story to create, It represents an activity associated with an object in the Asana system.
+ * @property {string} [story.text] - The plain text of the comment to add. Cannot be used with html_text.
```

Comment:
I don't think you need the object name in `[story.text]` though. Look at HTTP, where we just list the property name.

I also don't think we should put all these in square brackets. In an object like this I'd assume options are optional unless otherwise stated. Also, nowhere in the docs do we really explain that `[x]` means "optional", so it's not a terrible useful notation.

Reply:
I have made some improvements and remove `params.story` in `createTaskStory()`.
I have grouped all params for story in a variable called story like this `const story = { text, html_text, is_pinned, sticker_name };`.

To create a comment using text 
`createTaskStory("128320", {text: "my comment"})`


To create a comment using html format text  
`createTaskStory("128320", {text: "<body>my comment</body>"})`

Reply:
Looks good to me!